sound/soc/sof/intel/hda-stream.c

```diff
@@ -122,19 +137,28 @@ int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
 	ioc = sdev->hda->no_ipc_position ?
 	      !stream->no_period_wakeup : 0;
 
-	for (i = 0; i < periods; i++) {
-		if (i == (periods - 1) && remain)
-			/* set the last small entry */
-			offset = hda_setup_bdle(sdev, dmab,
-						stream, &bdl, offset,
-						remain, 0);
+	for (i = 0; i < entries; i++) {
+		bytes = period_bytes - offset % period_bytes;
+		if (is_wrap)
+			bytes = min(bytes, init_offset - offset);
 		else
-			offset = hda_setup_bdle(sdev, dmab,
-						stream, &bdl, offset,
-						period_bytes, ioc);
+			bytes = min(bytes, stream->bufsize - offset);
```

Comment:
@keyonjie I feel like the logic might be wrong here. For the normal stream case where offset is 0, we will end up with "entries" number of buffers of size period_bytes. But the last fragment might not be a full period.

Reply:
@keyonjie sorry this part is fine.